packages/contracts/src/interest/MultiTokenVault.sol

```diff
@@ -1,232 +1,280 @@
+
+        _deposit(assets, receiver, _msgSender(), depositPeriod, shares);
     }
 
-    // =============== Deposit ===============
+    /**
+     * @inheritdoc IMultiTokenVault
+     */
+    function redeemForDepositPeriod(
+        uint256 shares,
+        address receiver,
+        address owner,
+        uint256 depositPeriod,
+        uint256 redeemPeriod
+    ) public virtual nonReentrant returns (uint256 assets) {
+        if (depositPeriod > redeemPeriod) {
+            revert MultiTokenVault__RedeemBeforeDeposit(owner, depositPeriod, redeemPeriod);
+        }
+
+        if (currentTimePeriodsElapsed() < redeemPeriod) {
+            revert MultiTokenVault__RedeemTimePeriodNotSupported(owner, currentTimePeriodsElapsed(), redeemPeriod);
+        }
+
+        uint256 maxShares = maxRedeemAtPeriod(owner, depositPeriod);
+
+        if (shares > maxShares) {
+            revert MultiTokenVault__ExceededMaxRedeem(owner, depositPeriod, shares, maxShares);
+        }
+
+        assets = previewRedeemForDepositPeriod(shares, depositPeriod, redeemPeriod);
+
+        _withdraw(shares, receiver, owner, _msgSender(), depositPeriod, assets);
+    }
 
     /**
-     * @dev See {IMultiTokenVault-convertToSharesForDepositPeriod}
+     * @inheritdoc IMultiTokenVault
      */
-    function convertToSharesForDepositPeriod(uint256 assets, uint256 depositPeriod)
+    function redeemForDepositPeriod(uint256 shares, address receiver, address owner, uint256 depositPeriod)
         public
-        view
         virtual
-        returns (uint256 shares);
+        returns (uint256)
+    {
+        return redeemForDepositPeriod(shares, receiver, owner, depositPeriod, currentTimePeriodsElapsed());
+    }
 
     /**
-     * @dev See {IMultiTokenVault-convertToShares}
+     * @inheritdoc IMultiTokenVault
      */
-    function convertToShares(uint256 assets) public view override returns (uint256 shares) {
-        return convertToSharesForDepositPeriod(assets, currentPeriodElapsed);
+    function setCurrentTimePeriodsElapsed(uint256 currentTimePeriodsElapsed_) public virtual onlyOwner {
+        _currentTimePeriodsElapsed = currentTimePeriodsElapsed_;
     }
 
     /**
-     * @dev See {IMultiTokenVault-previewDeposit}
+     * @inheritdoc IMultiTokenVault
      */
-    function previewDeposit(uint256 assets) public view override returns (uint256 shares) {
-        return convertToShares(assets);
+    function asset() public view virtual returns (address) {
+        return address(ASSET);
     }
 
     /**
-     * @dev See {IMultiTokenVault-deposit}
+     * @inheritdoc IMultiTokenVault
      */
-    function deposit(uint256 assets, address receiver) public virtual override returns (uint256) {
-        uint256 shares = previewDeposit(assets);
-
-        ASSET.safeTransferFrom(_msgSender(), address(this), assets);
-
-        _mint(receiver, currentPeriodElapsed, shares, "");
+    function totalAssets() public view returns (uint256) {
```

Comment:
this will only be correct while assets are held by MultiTokenVault.  we should make this virtual as a child contract in RWA (such as our Product) will transfer the USDC out of the contract.